### ISSUE TITLE
Using Startup class not from the execution Assembly

### ISSUE THREAD

**Issue by YogiBear52** [aspnet/AspNetCore#7315]
### Describe the bug
Hey,

I am trying to use a Startup class like that
`WebHost.CreateDefaultBuilder(args).UseStartup<Startup>().Build().Run()`
When the Startup class is located in the execution assembly everything works fine.
When the Startup class is located in a different assembly, the service runs but things go wrong during requests - the service throws a 404 without any explanation.

Why is it important for me to put my Startup class in a different assembly?
Because I have many services that are using the exact same Startup configuration and I want to prevent duplication.

### To Reproduce
1. Creating new template of Asp.net core 2.2 WebApi
2. Request is working
3. Creating new Project of .Net Standard 2.0
4. Moved the Startup class from the Asp.net Core project to the .Net Standard Project
5. Added these two nugets to the .Net Standard project : 'Microsoft.AspNetCore.Mvc' and 'Microsoft.AspNetCore.Diagnostics' Nugets.
6. Add Reference from the Asp.net core project the .Net Standard project
7. Request is not working

Thank you :)

**Comment by anurse** [comment]
Can you share a runnable sample that reproduces the problem? That way we get an exact repro and can investigate in depth.

**Comment by davidfowl** [comment]
It might be worth looking at this issue again but I don't know if we'll have time to fix it. The workaround is to do something like this:

```C#
public class Program
{
    public static void Main(string[] args)
    {
        CreateWebHostBuilder(args).Build().Run();
    }

    public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
        WebHost.CreateDefaultBuilder(args)
            .UseStartup<Startup>()
            // Ignore the startup class assembly as the "entry point" and instead point it to this app
            .UseSetting(WebHostDefaults.ApplicationKey, typeof(Program).GetTypeInfo().Assembly.FullName); 
}
```

We've had this come up a number of times and it might make sense to handle this somehow.

https://github.com/aspnet/Hosting/issues?q=is%3Aissue+ApplicationKey+is%3Aclosed

**Comment by davidfowl** [comment]
We could clean this up a bit by adding a `UseEntryPoint` method that sets the entry point assembly. Another thing we should consider doing is splitting application name and entrypoint (but this would be a breaking change for people that originally worked around this issue via the above). The application entry point is used by MVC today to figure out where controllers are so it doubles as both the "application name" and the "entrypoint" assembly.

**Comment by anurse** [comment]
Seems reasonable. Any reason we can't default the entry point to the assembly containing the Startup type? Also the name `UseEntryPoint` is a bit overloaded (especially with some of the generic host stuff I've talked about with @glennc). Perhaps `UsePrimaryAssembly` or `UseApplicationAssembly`?

Also, does MVC have options to add/change the assembly that's scanned?

**Comment by davidfowl** [comment]
> Any reason we can't default the entry point to the assembly containing the Startup type? 

We do and that's the bug. The Startup class is in the class library and controllers are in the main application.

> Also the name UseEntryPoint is a bit overloaded (especially with some of the generic host stuff I've talked about with @glennc). Perhaps UsePrimaryAssembly or UseApplicationAssembly?

Yea, lets bikeshed the name. @khellang Proposed a PR a while back with similar functionality. We could even decide that this doesn't belong in hosting.

> Also, does MVC have options to add/change the assembly that's scanned?

Not the primary assembly. This happens very early (during DI setup) https://github.com/aspnet/AspNetCore/blob/258d34e3828a1870a16d13cd3c62d1b7a65acc4a/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L81-L88

**Comment by anurse** [comment]
> We do and that's the bug.

Gotcha, just re-read the issue.

**Comment by khellang** [comment]
What makes it super confusing is that it's called `ApplicationName` with no reference to assemblies.

I remember a couple of issues where people got tripped up because things stopped working by just changing the application name. Giving an instance of your app a pet name is a perfectly valid scenario, and so is changing the "entry point assembly" (not to be confused with an IL entrypoint). I think both should be allowed.

It's also kinda ironic that we're on the third breaking iteration of the IHostingEnvironment/IWebHostEnvironment and haven't used the opportunity to correct the naming mistake 😅

**Comment by khellang** [comment]
And, for reference: https://github.com/aspnet/Hosting/pull/1180. It 3.0 season - time to break someone!

**Comment by davidfowl** [comment]
> It's also kinda ironic that we're on the third breaking iteration of the IHostingEnvironment/IWebHostEnvironment and haven't used the opportunity to correct the naming mistake 😅

We're not though. We haven't broken anything until now. We added new interfaces and it's time to add the new property while we still can (during 3.0).

**Comment by davidalpert** [comment]
I'm confused why dotnet/aspnetcore#14183 was closed as "no work to do" - can anyone shed light on that decision? This is the closest issue I can find.

I spent some time digging through the code in this repo trying to figure out why my ASPNETCORE 3.x app was not respecting the `APPLICATION_NAME` set up in my app configuration and learned that using the `DefaultWebHostBuilder` and configuring a `Startup` class injects an additional configuration provider (last so that it _overrides_ my app configuration) and forces the app name to the AssemblyName of the DLL containing the `Startup` class.

I'm curious why it does not simply respect the `APPLICATION_NAME` as configured by `IConfigurationBuilder` instance returned from the `ConfigureAppConfiguration` method?

I'm using a Generic Host and then calling `ConfigureWebHostDefaults`, so maybe this is a holdover behavior in the Web Host from before the Generic Host was created?

It appears to place the [Generic Host guidance for setting `ApplicationName`](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-2.1#application-key-name) in conflict with the [Web Host guidance for setting `ApplicationName`](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-2.1#application-key-name)

Eventually I'm hoping that the responsibility of overriding the `ApplicationName` config setting can be removed from the web host responsibility and handled in a consistent way by the Generic Host

**Comment by hellfirehd** [comment]
What makes this worse is that the [Identity Templates](https://github.com/dotnet/aspnetcore/blob/master/src/Identity/UI/src/Areas/Identity/Pages/V4/_Layout.cshtml) are using `@Environment.ApplicationName`.  So the Title and Copyright are now the name of my assembly, not the actual Application Name that I want the public to see.

It's really a shame that this wasn't addressed in 3.0.  Maybe 5.0?

**Comment by alberto-chiesa** [comment]
Question (and proposal): why are we not using directly Assembly.GetEntryAssembly() as a first choice?

In my case (I don't use a startup class), I just put a line after calling Configure with:
```
        protected override IHostBuilder CreateHostBuilder()
            => new HostBuilder()
                .UseServiceProviderFactory(ctx => ServiceProviderFactory)
                .ConfigureWebHost(webHostBuilder => webHostBuilder
                    .UseKestrel(options => { })
                    .UseContentRoot(Directory.GetCurrentDirectory())
                    .ConfigureServices(ConfigureServices)
                    .Configure(ConfigureWebHost)
                    .UseSetting(WebHostDefaults.ApplicationKey, Assembly.GetEntryAssembly().FullName));
```
And everything works again.
Seems to me it's a pretty reasonable default.
Maybe it's not when a Framework launcher is used? I'm on Windows, and using self contained exe, so the whole "deploy a dll" is something I never played with.
Maybe a fallback logic could be in place.

**Comment by hillin** [comment]
> Question (and proposal): why are we not using directly Assembly.GetEntryAssembly() as a first choice?
> 
> In my case (I don't use a startup class), I just put a line after calling Configure with:
> 
> ```
>         protected override IHostBuilder CreateHostBuilder()
>             => new HostBuilder()
>                 .UseServiceProviderFactory(ctx => ServiceProviderFactory)
>                 .ConfigureWebHost(webHostBuilder => webHostBuilder
>                     .UseKestrel(options => { })
>                     .UseContentRoot(Directory.GetCurrentDirectory())
>                     .ConfigureServices(ConfigureServices)
>                     .Configure(ConfigureWebHost)
>                     .UseSetting(WebHostDefaults.ApplicationKey, Assembly.GetEntryAssembly().FullName));
> ```
> 
> And everything works again.
> Seems to me it's a pretty reasonable default.
> Maybe it's not when a Framework launcher is used? I'm on Windows, and using self contained exe, so the whole "deploy a dll" is something I never played with.
> Maybe a fallback logic could be in place.

Using `Assembly.GetEntryAssembly().FullName` does not work for us, but `Assembly.GetEntryAssembly().GetName().Name` does work. Thanks!

**Comment by davidfowl** [comment]
Entry assembly would break things like tests where the test process is the entry assembly and your code is hosted as a library inside of the test process.

**Comment by theo-albers** [comment]
Last week I spent some time preparing our .NET 5 solution for .NET 6. We have 3 API projects with similar base setup like Application Insights, Azure App Configuration, Masstransit, Azure BLOB Storage, OpenId Connect, Logging. To eliminate Sonar complaints about duplicate code, we moved some setup to a base class in a library project. In de .NET 5 setup I placed the code in a StartUp class in the API projects. 

When I prepare the code for .NET 6, I eliminated the StartUp class: I put everyting in a simple AppBase class. The IHost is constructed in the base class and we have some virtual members which can be overridden in the API projects. One project uses Razor views, since that's the Identity project. The other API projects are pure Controller projects. 

When swapping out the StartUp class I ran into the first problem: 404 on every REST call. Ah, since the AddControllers class is now in the library project, the controllers are no longer discovered. The solution was to place this in the library project, under the assumption that the extended App class would be in the API project:
```csharp
services.AddControllers().AddApplicationPart(GetType().Assembly);
```

Now I'm able to run the solution and everything works. I deploy the solution to Azure and try to login. Problem: 404 on every Razor page. When I look at the logs via Kudu in Azure at the Linux App Service, I see an exception:
```
Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware[1]
An unhandled exception has occurred while executing the request.
System.InvalidOperationException: The default Identity UI layout requires a partial view '_LoginPartial' usually located at '/Pages/_LoginPartial' or at '/Views/Shared/_LoginPartial' to work.
```

This error is misleading, but put me on the right track: I have no _LoginPartial in my code, so there must be something wrong with views. After investigating ASP.NET 5 code via GitHub I figured out the differences between my local environment and the production environment. Locally I have this setting in my launch settings:
```json
"ASPNETCORE_HOSTINGSTARTUPASSEMBLIES": "Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation"
``` 

The local environment finds the views, but deployed environment doesn't. I can fix this by explicitly loading the views via the following snippet:
```csharp
 public static IMvcCoreBuilder AddControllersAndRazorViewParts(this IMvcCoreBuilder me, Type type)
        {
            me.AddApplicationPart(type.Assembly);

            // Load the *.Views.dll compiled Razor assembly which is marked via the RelatedAssemblyAttribute in the main assembly.
            foreach (var relatedAssembly in RelatedAssemblyAttribute.GetRelatedAssemblies(type.Assembly, false))
            {
                me.AddApplicationPart(relatedAssembly);
            }

            return me;
        }
```
I invoke this in my AppBase class. Searching some more I discover that ApplicationName is used by AddMvcCore when it adds the PartManager. It implicitly loads the parts from the assembly inferred by the ApplicationName. I found more issues related to this stuff:
- https://github.com/dotnet/aspnetcore/issues/34914
- https://github.com/aspnet/Hosting/issues/1179

It's only *now*, after having hit this problem, I start reading the documentation and things start making sense:

https://docs.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-6.0?view=aspnetcore-6.0
"...
Razor compiler no longer produces a separate Views assembly
..."

Finally I find this gist https://gist.github.com/davidfowl/0e0372c3c1d895c3ce195ba983b1e03d

"...
The application name always defaults to the entry point assembly's name Assembly.GetEntryAssembly().GetName().FullName. When using the WebApplicationBuilder in a library, you will need to explicitly change the application name to the library's assembly to allow MVC's application part discovery to work (finding controllers, views etc) (see the Cheatsheet for instructions on how to do this).
..."

This old snippet is somewhat confusing https://github.com/aspnet/Mvc/issues/6685:
"...
The doc does needs an update. We should remove that configuring the application name to something arbitrary.
..."

Long story short: when removing the StartUp class in ASP.NET 5 to prepare for .NET 6 you hit runtime issues when you still use .NET 5. The documentation is only useful, after you solved the problems. On top of that .NET 6 has an optimization to no longer place Razor views in a separate View.dll. This makes it very hard to see why your .NET 5 code won't work.

**Comment by davidalpert** [comment]
> Entry assembly would break things like tests where the test process is the entry assembly and your code is hosted as a library inside of the test process.

What kind of a test might be broken by this scenario?

**Comment by akonyer** [comment]
I am running into this issue using the pact contract testing framework.

I am spinning up a separate web server for API contract testing, and I'm spinning it up from a unit testing assembly.

I spent a couple of days trying to figure out why I was only getting 404's when I wrap my test startup class in another one to add some middleware for testing.

@davidfowl's solution above helped me resolve this issue. Thanks a lot :)
